### PR TITLE
Settings constraints

### PR DESCRIPTION
**Pull request by vitlibar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category:
- New Feature
- Improvement

This PR makes possible to set constraints in `users.xml`, for example

```
<profiles>
  <user_profile>
    <max_memory_usage>10000000000</max_memory_usage>
    <force_index_by_date>0</force_index_by_date>
    ...
    <constraints>
      <max_memory_usage>
        <min>5000000000</min>
        <max>20000000000</max>
      </max_memory_usage>
      <force_index_by_date>
        <readonly/>
      </force_index_by_date>
    </constraints>
  </user_profile>
</profiles>
```

It means that `max_memory_usage` should be between `5000000000` and `20000000000`.
The following query will raise an exception:
```
SET max_memory_usage=20000000001
```

### CODE REVIEW

File: dbms/src/Parsers/ParserSetQuery.cpp [hunk 1]
```diff
@@ -31,6 +31,7 @@ static bool parseNameValuePair(ASTSetQuery::Change & change, IParser::Pos & pos,
     if (!value_p.parse(pos, value, expected))
         return false;
 
+    String change_name;
```
Comment (KochetovNicolai) [inline dbms/src/Parsers/ParserSetQuery.cpp]: Do we need this field?

File: dbms/src/Storages/MergeTree/MergeTreeSettings.h [hunk 2]
```diff
@@ -14,172 +21,166 @@ class ASTStorage;
...
-    /** Check delay of replicas settings. */                                                                  \
-                                                                                                              \
-    /** Period to check replication delay and compare with other replicas. */                                 \
-    M(SettingUInt64, check_delay_period, 60)                                                                  \
-                                                                                                              \
-    /** Period to clean old queue logs, blocks hashes and parts */                                            \
-    M(SettingUInt64, cleanup_delay_period, 30)                                                                \
-    /** Add uniformly distributed value from 0 to x seconds to cleanup_delay_period                           \
+    /** How many rows correspond to one primary key value. */                                                     \
+    M(SettingUInt64, index_granularity, 8192, "")                                                                 \
```
Comment (KochetovNicolai) [inline dbms/src/Storages/MergeTree/MergeTreeSettings.h]: It's better to move comments to setting description for every setting.

File: dbms/src/Core/SettingsCommon.h [hunk 3]
```diff
@@ -408,12 +430,389 @@ class SettingLogsLevel
...
+        size_t size() const { return infos.size(); }
+        const MemberInfo & operator[](size_t index) const { return infos[index]; }
+        const MemberInfo * begin() const { return infos.data(); }
+        const MemberInfo * end() const { return infos.data() + infos.size(); }
+
+        size_t findIndex(const StringRef & name) const
+        {
+            auto it = by_name_map.find(name);
+            if (it == by_name_map.end())
+                return static_cast<size_t>(-1);
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply dbms/src/Core/SettingsCommon.h]: comment added

File: dbms/src/Parsers/ParserSetQuery.cpp (hunk 1 above)
Comment (vitlibar) [reply dbms/src/Parsers/ParserSetQuery.cpp]: removed

File: dbms/src/Storages/MergeTree/MergeTreeSettings.h (hunk 2 above)
Comment (vitlibar) [reply dbms/src/Storages/MergeTree/MergeTreeSettings.h]: done

**Comment by alexey-milovidov** [comment]
read_only is inconsistent with readonly

**Comment by vitlibar** [comment]
> read_only is inconsistent with readonly

There is no inconsistency. Both setting and constraint are named `readonly`.

File: dbms/tests/integration/helpers/client.py [hunk 4]
```diff
@@ -89,3 +93,20 @@ def get_answer(self):
...
+        self.stderr_file.seek(0)
+
+        stdout = self.stdout_file.read()
+        stderr = self.stderr_file.read()
+
+        if self.timer is not None and not self.process_finished_before_timeout and not self.ignore_error:
+            raise QueryTimeoutExceedException('Client timed out!')
+
+        if (self.process.returncode == 0):
+            raise QueryRuntimeException('Client expected to be failed but was succeeded! stdout: {}'.format(stdout))
```
Comment (kagan770) [inline dbms/tests/integration/helpers/client.py]: "was" is not grammatically correct and should be omitted.  Would you please correct to the following:
"Client expected to be failed but succeeded"

**Comment by kagan770** [comment]
<max_memory_usage> - can that tag be changed to <memory_usage>  - that's due to the fact that min AND max can be set in such XML block.  Also - should we check for min values and if set for less than that the user will receive the error similar to "max" constraint?

**Comment by vitlibar** [comment]
> <max_memory_usage> - can that tag be changed to <memory_usage> - that's due to the fact that min AND max can be set in such XML block. Also - should we check for min values and if set for less than that the user will receive the error similar to "max" constraint?

`max_memory_usage` is a name of a setting. The settings are listed in the [documentation](https://clickhouse.yandex/docs/en/operations/settings/settings/), and `max_memory_usage` means [the maximum memory usage for processing of a single query](https://github.com/yandex/ClickHouse/blob/4c7d71af1c9233958090a9e18206456c3f842894/dbms/src/Core/Settings.h#L287). Most of the settings can be changed in SQL script like this:

`
SET max_memory_usage = 100000000;
`

Here in this PR I implemented per-user constraints which can be used for server's administrators to prohibit some users from changing some of the settings. I referenced to `max_memory_usage` here just as an example, the constraints can be applied to any settings.

File: dbms/tests/integration/helpers/client.py (hunk 4 above)
Comment (vitlibar) [reply dbms/tests/integration/helpers/client.py]: Sure. https://github.com/yandex/ClickHouse/pull/5573

**Comment by vitlibar** [comment]
About violation of the lower boundary: for the constraints from my example the following query raises an exception too:

```
SET max_memory_usage=4999999999
```

**Pull request merged**